pkg/kubevirt/utils.go

```diff
@@ -171,3 +162,68 @@ func nodeRole(ctx *context.MachineContext) string {
 	}
 	return constants.WorkerNodeRoleValue
 }
+
+func cloudinitVolume(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec, ctx *context.MachineContext, preferredType string) kubevirtv1.Volume {
+	switch preferredType {
+	case "CloudInitNoCloud":
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitNoCloud: &kubevirtv1.CloudInitNoCloudSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
+				},
+			},
+		}
+	default:
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
+				},
+			},
+		}
+	}
+}
+
+func cloudinitDisk(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec, preferredType string) kubevirtv1.Disk {
+	switch preferredType {
+	case "CloudInitNoCloud":
+		return kubevirtv1.Disk{
+			Name: cloudInitVolumeName,
+			DiskDevice: kubevirtv1.DiskDevice{
+				CDRom: &kubevirtv1.CDRomTarget{
+					Bus: kubevirtv1.DiskBusSATA,
+					ReadOnly: func() *bool {
+						b := true
+						return &b
+					}(),
+				},
```

Comment:
To some tests the CDROM type appears to work better with Talos

Reply:
How about we default to the `disk` approach that we already have today, but only create that disk if there's not already a disk with the name `cloudinitvolume`

So this would allow you to create a disk however you want and our controllers won't touch it. But if the cloudinit disk doesn't exist, our controllers will give one as a default.

Reply:
I tried it with Disk again and it appears to work fine.
I like the flow of providing a disk/volume with a given name and it fills in the rest using your changes on top.

Made it so it adds to if it doesn't already exist 57d40f5